source/mesh_refinement/thermal_energy_density.cc

```diff
@@ -101,11 +101,36 @@ namespace aspect
                                       this->get_mpi_communicator());
       vec = vec_distributed;
 
-      DerivativeApproximation::approximate_gradient  (this->get_mapping(),
-                                                      this->get_dof_handler(),
-                                                      vec,
-                                                      indicators,
-                                                      this->introspection().component_indices.temperature);
+      try
+        {
+          DerivativeApproximation::approximate_gradient  (this->get_mapping(),
+                                                          this->get_dof_handler(),
+                                                          vec,
+                                                          indicators,
+                                                          this->introspection().component_indices.temperature);
+        }
+      catch (std::exception &exc)
+        {
+          std::cerr << std::endl << std::endl
+                    << "----------------------------------------------------"
+                    << std::endl;
+          std::cerr << "Exception on MPI process <"
+                    << Utilities::MPI::this_mpi_process(MPI_COMM_WORLD)
+                    << "> while running mesh refinement plugin <"
+                    << "thermal energy density"
+                    << ">: " << std::endl
+                    << "The mesh refinement plugin failed to compute a thermal energy density gradient. "
+                    << "This could be caused by having a constant temperature in the model. If this is "
+                    << "the case, please set an other mesh refinement strategy. DEAL.II error:"
```

Comment:
Is that really the exception you get? I would think that that might happen because you have too few cells (no neighbors in all coordinate directions) to compute a finite difference approximation. The solution *happens* to be constant if you have constant boundary values, but that's a coincidence.

Reply:
```suggestion
                    << "the case, please set an other mesh refinement strategy. deal.II error:"
```